src/decoder/mod.rs

```diff
@@ -536,20 +536,16 @@ impl<R: Read> Reader<R> {
         self.data_stream.clear();
         self.current_start = 0;
         self.prev_start = 0;
-        let width = self.info().width;
         if self.info().interlaced {
-            while let Some(InterlacedRow {
-                data: row,
-                interlace,
-                ..
-            }) = self.next_interlaced_row()?
-            {
+            let width = self.info().width;
+            let mut row = vec![0; self.output_line_size(width)];
```

Comment:
Is this additional allocation benchmarked anywhere? And from the looks below, maybe it could remain unchanged with the code in `next_interlaced_row` being preserved?

Reply:
Before this commit the following allocations happen when handling interlaced images:

1. `self.scratch_buffer` is taken and [resized](https://github.com/image-rs/image-png/blob/4d5a4b1f2667e69a4f0c2db843098599d3e88e6d/src/decoder/mod.rs#L612) via [`output_buffer.resize(output_line_size, ...)`] once per call to `next_row` / `next_interlaced_row`.  For interlaced images `output_line_size` can grow with each interlacing pass and so potentially there are 7 (re)allocations per image frame.
2. There is no `row` vector allocated inside `fn next_frame`

After this commit:
1. `self.scratch_buffer` is never allocated when going via `next_frame` (`Vec::new()` [here](https://github.com/image-rs/image-png/blob/4d5a4b1f2667e69a4f0c2db843098599d3e88e6d/src/decoder/mod.rs#L233) doesn't allocate)
2. `next_frame` will allocate the `row` vector, as you've highlighted above.

So, when decoding interlaced images:
- Before this commit: 1 call to `next_frame` can result in 1-7 allocations of row storage
- After this commit: 1 call to `next_frame` always results in 1 allocation of row storage

-------------------------

I note that `next_frame` only goes through `read_row` (or `next_interlaced_row` before this commit) for _interlaced_ images.  Nevertheless `next_row` / `read_row` are useful not just for interlaced images, but for any user/caller who wants 1) incremental decoding (e.g. wants decoded rows as soon as they are available in the input) and/or 2) wants to post-process rows (e.g. RGB => RGBA, or maybe gamma correction) while the row is still hot in L1 cache.  So the performance of `read_row` matters not only for interlaced images, but for all images.  (Incremental decoding is tricky and may need additional PRs to expose the currently-internal-only     `UnexpectedEof` and `UnexpectedEndOfChunk` errors.)

-------------------------

> Is this additional allocation benchmarked anywhere?

No, I haven't specifically benchmarked the impact of this allocation (other than executing end-to-end decoding benchmarks).

> And from the looks below, maybe it could remain unchanged with the code in next_interlaced_row being preserved?

I am not sure if I understand your proposal.  We can't decode the interlaced row into the caller/user-provided `buf`, because the decoded interlaced row may be "squished" (in the first 6 out of 7 adam7 passes) and needs to be "splat" into the final image `buf` via `adam7::expand_pass`.  In theory during pass 7 we could indeed decode directly into the user-provided buffer, but this seems complicated (and interlaced images are not that common anyway - only around 1% of images on top-500 websites - see my year-old corpus experiments [here](https://docs.google.com/spreadsheets/d/1CEgr4ttd2agOZi-MFOjqJGqfDVFUKDp-zKbrF-8Xwyk/edit?gid=1170535878#gid=1170535878)).

Reply:
After thinking a little bit more about this feedback I now see that the changes in `fn next_frame` are somewhat separate and independent from 1) introducing `read_row` and 2) using `read_row` 2a) from `next_interlaced_row` and 2b) from the new benchmark.  And I expect this part of the changes to be mostly performance-neutral.  So let me revert this part of the changes.

Reverting this part of the changes will also let me avoid merge conflicts as I try to experiment with figuring out the best shape of a public API for exposing `adam7::expand_pass`.  (I've initially planned to implement `Reader.expand_interlaced_row(dst, row, interlace_info, option_to_override_bytes_per_pixel)` but realized that hard-coding using image width won't work if some APNG frames use smaller width.  Now I lean toward exposing `InterlaceInfo.dest_row_number() -> u32-or-usize` and `InterlaceInfo.expand(expanded_row, interlaced_row, bytes_per_pixel, maybe_no_need_to_pass_width_herre)`.  I think that `dest_row_number` will be helpful for finding the target row bytes regardless of whether the client works with image-sized buffers or frame-sized buffers.)

Reply:
>  [...] let me revert this part of the changes.

Done.

Reply:
Thanks for experimenting with different combinations, feel free to choose the best method based on those numbers. I don't feel very strongly about the interlaced case. One allocation per image isn't so much of a problem (if the numbers don't show it). Just one correct: `resize` does _at most_ one allocation, not exactly one. This should be notable for APNG where later frames can re-use the same buffer for in subsequence `fdAT` chunks and re-use the same interlacing.